sound/soc/sof/ipc4-topology.c

```diff
@@ -0,0 +1,678 @@
+			 fmt->ch_cfg,
+			 fmt->interleaving_style,
+			 fmt->fmt_cfg);
+	}
+}
+
+static int sof_ipc4_get_audio_fmt(struct snd_soc_component *scomp,
+				  struct snd_sof_widget *swidget,
+				  struct sof_ipc4_available_audio_format *available_fmt,
+				  bool has_out_format)
+{
+	struct sof_ipc4_base_module_cfg *base_config;
+	struct sof_ipc4_audio_format *out_format;
+	int audio_fmt_num = 0;
+	int ret;
+
+	ret = sof_update_ipc_object(scomp, &audio_fmt_num,
+				    SOF_IPC4_AUDIO_FMT_NUM_TOKENS, swidget->tuples,
+				    swidget->num_tuples, sizeof(audio_fmt_num), 1);
+
+	if (ret != 0 || audio_fmt_num <= 0) {
+		dev_err(scomp->dev, "Invalid number of audio formats: %d\n", audio_fmt_num);
+		return -EINVAL;
+	}
+	available_fmt->audio_fmt_num = audio_fmt_num;
+
+	base_config = kcalloc(available_fmt->audio_fmt_num, sizeof(*base_config), GFP_KERNEL);
+	if (!base_config)
+		return -ENOMEM;
+
+	ret = sof_update_ipc_object(scomp, base_config,
+				    SOF_IPC4_COMP_TOKENS, swidget->tuples,
+				    swidget->num_tuples, sizeof(*base_config),
+				    available_fmt->audio_fmt_num);
+
+	if (ret != 0) {
+		dev_err(scomp->dev, "parse base_config tokens failed %d\n", ret);
+		goto err_in;
+	}
+
+	ret = sof_update_ipc_object(scomp, &base_config->audio_fmt,
+				    SOF_IPC4_IN_AUDIO_FORMAT_TOKENS, swidget->tuples,
+				    swidget->num_tuples, sizeof(*base_config),
+				    available_fmt->audio_fmt_num);
+
+	if (ret != 0) {
+		dev_err(scomp->dev, "parse base_config audio_fmt tokens failed %d\n", ret);
+		goto err_in;
+	}
+
+	sof_ipc4_dbg_audio_format(scomp->dev, &base_config->audio_fmt,
+				  sizeof(*base_config),
+				  available_fmt->audio_fmt_num,
+				  swidget->widget->name, __func__);
+
+	available_fmt->base_config = base_config;
+
+	if (!has_out_format)
+		goto finish;
+
+	out_format = kcalloc(available_fmt->audio_fmt_num, sizeof(*out_format), GFP_KERNEL);
+	if (!out_format) {
+		ret = -ENOMEM;
+		goto err_in;
+	}
+
+	ret = sof_update_ipc_object(scomp, out_format,
+				    SOF_IPC4_OUT_AUDIO_FORMAT_TOKENS, swidget->tuples,
+				    swidget->num_tuples, sizeof(*out_format),
+				    available_fmt->audio_fmt_num);
+
+	if (ret != 0) {
+		dev_err(scomp->dev, "parse output audio_fmt tokens failed\n");
+		goto err_out;
+	}
+
+	available_fmt->out_audio_fmt = out_format;
+	sof_ipc4_dbg_audio_format(scomp->dev, out_format, sizeof(*out_format),
+				  available_fmt->audio_fmt_num,
+				  swidget->widget->name, __func__);
+
+finish:
+	return 0;
+
+err_out:
+	kfree(out_format);
+err_in:
+	kfree(base_config);
+
+	return ret;
+}
+
+
+static void sof_ipc4_widget_free_comp(struct snd_sof_widget *swidget)
+{
+	kfree(swidget->private);
+}
+
+static int sof_ipc4_widget_setup_pcm(struct snd_sof_widget *swidget)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct sof_ipc4_host *host;
+	size_t size = sizeof(*host);
+	int node_type = 0;
+	int ret;
+
+	host = (struct sof_ipc4_host *)
+	       sof_ipc4_comp_alloc(swidget, &size, swidget->pipeline_id);
```

Comment:
this seems wrong, sof_ipc4_comp_alloc return a sof_ipc_comp pointer which is IPC3 specific

Reply:
anyway once you fix this, you can remove the casting because the function returns a void *

Reply:
I use `host = kzalloc(sizeof(*host), GFP_KERNEL);` instead